code/obj/item/pet_carrier.dm

```diff
@@ -0,0 +1,321 @@
+		src.carrier_occupants.Add(mob_to_add)
+		src.vis_contents_proxy.vis_contents.Add(mob_to_add)
+		src.UpdateIcon()
+
+	/// Directly ejects a target mob from the carrier.
+	proc/eject_mob(mob/mob_to_eject)
+		if (!mob_to_eject)
+			return
+		MOVE_OUT_TO_TURF_SAFE(mob_to_eject, src)
+		mob_to_eject.vis_flags &= ~src.mob_vis_flags
+		src.carrier_occupants.Remove(mob_to_eject)
+		src.vis_contents_proxy.vis_contents.Remove(mob_to_eject)
+		src.UpdateIcon()
+
+	/// Checks humans and mob critters for any empty hands.
+	proc/find_empty_hand(mob/living/mob_to_check)
+		if (ishuman(mob_to_check))
+			var/mob/living/carbon/human/human_to_check = mob_to_check
+			if (human_to_check.hand && !human_to_check.r_hand) return TRUE
+			else if (!human_to_check.l_hand) return TRUE
+		if (ismobcritter(mob_to_check))
+			var/mob/living/critter/mob_critter_to_check = mob_to_check
+			for (var/datum/handHolder/current_hand in mob_critter_to_check.hands)
+				if (!current_hand.item) return TRUE
+		return FALSE
+
+	/// Deals damage to the door. If the remaining health <= 0, release everyone and reset the carrier.
+	proc/take_door_damage(damage)
+		src.door_health -= damage
+		if (src.door_health <= 0)
+			for (var/occupant in src.carrier_occupants)
+				src.eject_mob(occupant)
+			src.visible_message("<span class='alert'>The door on [src] busts wide open, releasing its occupants!</span>")
+			src.door_health = src.door_health_max
+		else
+			src.visible_message("<span class='alert'>The door on [src] rattles!</span>")
+
+	/// Calls src.AttackSelf(user) with a context action. Yeah, I know.
+	verb/release_occupant_verb(mob/user)
+		set name = "Release occupant"
+		set category = "Local"
+		set src in oview(1)
+
+		if (can_act(user))
+			return
+
+		src.AttackSelf(user)
+
+/obj/item/pet_carrier/admin_crimes
+	name = "pet carrier (ADMIN CRIMES EDITION)"
+	desc = "A surprisingly roomy carrier for transporting living things. All of them."
+	allowed_mob_type = /mob
+	can_break_out = FALSE
+	carrier_max_capacity = INFINITY
+	door_health_max = INFINITY
+
+/// Pertains to actions executed by the pet carrier.
+/datum/action/bar/icon/pet_carrier
+	duration = 2 SECONDS
+	interrupt_flags = INTERRUPT_MOVE | INTERRUPT_STUNNED | INTERRUPT_ATTACKED
+	var/mob/mob_owner
+	var/mob/target
+	var/obj/item/pet_carrier/carrier
+	var/action
+
+	New(mob/target, obj/item/pet_carrier/item, icon, icon_state, carrier_action)
+		..()
+		src.target = target
+		if (istype(item, /obj/item/pet_carrier))
+			src.carrier = item
+		else
+			logTheThing(LOG_DEBUG, src, "/datum/action/bar/icon/pet_carrier called with invalid type [item?.type].")
+		src.icon = icon
+		src.icon_state = icon_state
+		src.action = carrier_action
+
+	onStart()
+		if (!ismob(owner))
+			interrupt(INTERRUPT_ALWAYS)
+			return
+		src.mob_owner = owner
+		if (BOUNDS_DIST(mob_owner, target) > 0 || !target || !mob_owner || !src.carrier || (src.action == TRAP_MOB && mob_owner.equipped() != carrier) || !carrier.find_empty_hand(mob_owner))
```

Comment:
I know action bars tend to be copy pasted to hell, but please just make this a proc